src/constraints.jl

```diff
@@ -0,0 +1,207 @@
+#############################################################################
+# AbstractShape
+# Abstract base type for the shape of constraints
+"""
+    AbstractShape
+
+Abstract vectorizable shape. Given a flat vector form of an object of shape
+`shape`, the original object can be obtained by [`reshape`](@ref).
+"""
+abstract type AbstractShape end
+
+"""
+    dual_shape(shape::AbstractShape)::AbstractShape
+
+Returns the shape of the dual space of the space of objects of shape `shape`. By
+default, the `dual_shape` of a shape is itself. See the examples section below
+for an example for which this is not the case.
+
+## Examples
+
+Consider polynomial constraints for which the dual is moment constraints and
+moment constraints for which the dual is polynomial constraints. Shapes for
+polynomials can be defined as follows:
+```julia
+struct Polynomial
+    coefficients::Vector{Float64}
+    monomials::Vector{Monomial}
+end
+struct PolynomialShape <: JuMP.AbstractShape
+    monomials::Vector{Monomial}
+end
+JuMP.reshape(x::Vector, shape::PolynomialShape) = Polynomial(x, shape.monomials)
+```
+and a shape for moments can be defined as follows:
+```julia
+struct Moments
+    coefficients::Vector{Float64}
+    monomials::Vector{Monomial}
+end
+struct MomentsShape <: JuMP.AbstractShape
+    monomials::Vector{Monomial}
+end
+JuMP.reshape(x::Vector, shape::MomentsShape) = Moments(x, shape.monomials)
+```
+The `dual_shape` allows to define the shape of the dual of polynomial and moment
+constraints:
+```julia
+dual_shape(shape::PolynomialShape) = MomentsShape(shape.monomials)
+dual_shape(shape::MomentsShape) = PolynomialShape(shape.monomials)
+```
+"""
+dual_shape(shape::AbstractShape) = shape
+
+"""
+    reshape(vectorized_form::Vector, shape::AbstractShape)
+
+Return an object in it original shape `shape` given its vectorized form
+`vectorized_form`.
+
+## Examples
+
+Given a [`SymmetricMatrixShape`](@ref) of vectorized form `[1, 2, 3]`, the
+following code retrieve the matrix `Symmetric(Matrix[1 2; 2 3])`:
+```julia
+reshape([1, 2, 3], SymmetricMatrixShape(2))
+```
+"""
+function reshape end
```

Comment:
Maybe stupid question: Given there is a Base.reshape, does this squash the Base definition in the scope of JuMP from this point on? What does it do to the docstrings?

Reply:
It defines a JuMP.reshape which is a different function form `Base.reshape` so `reshape` in the JuMP scope and one need to explicitely write `Base.reshape`